ryan_savill/principal_feature_analysis/readme.md

```diff
@@ -0,0 +1,902 @@
+      <td>0</td>
+    </tr>
+    <tr>
+      <th>image_stdev</th>
+      <td>3</td>
+    </tr>
+    <tr>
+      <th>avg distance of 2 closest points</th>
+      <td>2</td>
+    </tr>
+    <tr>
+      <th>avg distance of 3 closest points</th>
+      <td>2</td>
+    </tr>
+    <tr>
+      <th>avg distance of 4 closest points</th>
+      <td>2</td>
+    </tr>
+    <tr>
+      <th>touching neighbor count</th>
+      <td>0</td>
+    </tr>
+  </tbody>
+</table>
+</div>
+
+
+
+What can we see here? Well each feature is now part of a cluster (numbers 0 - 6, since we have chosen 7 clusters). What we can also see is that quite a few features are part of the same cluster( = 2), while some features are only part of a unique cluster without any other features. This means that these features do not correlate (much at least) with any other features and will be chosen by our feature analysis. The clusters which have 2 or more features as part of it are the clusters where a feature has to be chosen. This is decided by the distances the features have from the cluster centres, so let's determine the distances from the cluster centres for each feature:
+
+
+```python
+# initialising a dictionary to keep the distances in:
+dists = defaultdict(list)
+
+# iterating through each feature (i) and it's cluster (c)
+for i, c in enumerate(clusters):
+    
+    # measuring the distance of each feature in A_q ([A_q[i, :]])
+    # and the cluster centers [cluster_centers[c, :]]
+    dist = euclidean_distances([A_q[i, :]], [cluster_centers[c, :]])[0][0]
+    
+    # each cluster is a key in the dictionary and the distance of each feature
+    # belonging to that cluster is saved under this key as a tupule (feature_idx, distance)
+    dists[c].append((i, dist))
+
+dists
+```
+
+
+
+
+    defaultdict(list,
+                {2: [(0, 0.039527541499107074),
+                  (4, 0.22202861661398224),
+                  (5, 0.1470198671974888),
+                  (6, 0.13180725994276707),
+                  (12, 0.09346378648078914),
+                  (13, 0.06610212866152891),
+                  (14, 0.04944098356286365)],
+                 5: [(1, 0.0)],
+                 6: [(2, 0.0)],
+                 1: [(3, 0.0)],
+                 4: [(7, 0.0)],
+                 0: [(8, 0.217150845350435),
+                  (10, 0.2117088702247021),
+                  (15, 0.29315857787851873)],
+                 3: [(9, 0.21819421392343466), (11, 0.21819421392343452)]})
+
+
+
+In our dists dictionary eah cluster has a list of tupules associated with it. Each tuple contains:
+(feature index, distance to cluster center)
+We can see that for the clusters with only one tupule associated (5,6,1,4), the features all have distance 0 since the cluster center IS the feature. For all other clusters we still have to determine which feature is closest to the cluster center. Once the features have been determined we can transform our original dataset to only include these features:
```

Comment:
```suggestion
We can see that for the clusters with only one tuple associated `(5,6,1,4)`, the features all have distance 0 since the cluster center IS the feature. For all other clusters we still have to determine which feature is closest to the cluster center. Once the features have been determined we can transform our original dataset to only include these features:
```